core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -615,7 +615,7 @@ class TransactionCoordinator(brokerId: Int,
     info("Shutting down.")
     isActive.set(false)
     scheduler.shutdown()
-    producerIdManager.shutdown()
+    producerIdGenerator.shutdown()
```

Comment:
Since we're externalizing this trait (`ProducerIdGenerator`), it seems odd that TransactionCoordinator takes it as a dependency, but then later assumes the responsibility of closing it. I think we might need to move the shutdown call to the same place we create the instance, just so the ownership is clear.

Alternatively (and maybe preferably?) we can keep the ProducerIdGenerator construction in the companion object and add a second `apply` method when we add the KIP-500 implementation.

Reply:
Either way, the new trait looks good

Reply:
Yeah, I agree that the shutdown of producer id generator should move to `KafkaServer` if it's where we create it. Alternatively, `TransactionCoordinator` could take a factory (i.e. () => ProducerIdGenerator) for creating a `ProducerIdGenerator` and keep lifecycle management in `TransactionCoordinator`.

Reply:
I think the factory with the `apply` method would work reasonably well.

Reply:
It felt weird to create the instance in `apply()` and hand it off to `TransactionCoordinator`, so I passed the factory and had the coordinator generate the instance.